engine/scene/sceneloader.cpp

```diff
@@ -0,0 +1,198 @@
+#include "sceneloader.h"
+
+#include "ball.h"
+#include "button.h"
+#include "checkbox.h"
+#include "colliderbox.h"
+#include "componentdeserializer.h"
+#include "config.h"
+#include "gameobject.h"
+#include "logger.h"
+#include "mesh.h"
+#include "player.h"
+#include "opponent.h"
+#include "scorearea.h"
+#include "scorecontroller.h"
+#include "settingsscreencontroller.h"
+#include "slider.h"
+#include "text.h"
+#include "titlescreencontroller.h"
+#include "transform.h"
+
+#include <nlohmann/json.hpp>
+
+#include <chrono>
+#include <fstream>
+
+namespace pong
+{
+
+SceneLoader::SceneLoader()
+{
+    // Engine components
+    mComponentMappings["Transform"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Transform>();
+    };
+
+    mComponentMappings["Mesh"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Mesh>();
+    };
+
+    mComponentMappings["ColliderBox"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<ColliderBox>();
+    };
+
+    mComponentMappings["Slider"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Slider>();
+    };
+
+    mComponentMappings["CheckBox"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<CheckBox>();
+    };
+
+    mComponentMappings["Button"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Button>();
+    };
+
+    mComponentMappings["Text"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Text>();
+    };
+
+    // User defined components
+    mComponentMappings["Player"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Player>();
+    };
+
+    mComponentMappings["Opponent"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Opponent>();
+    };
+
+    mComponentMappings["Ball"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<Ball>();
+    };
+
+    mComponentMappings["ScoreArea"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<ScoreArea>();
+    };
+
+    mComponentMappings["ScoreController"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<ScoreController>();
+    };
+
+    mComponentMappings["TitleScreenController"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<TitleScreenController>();
+    };
+
+    mComponentMappings["SettingsScreenController"] = [](GameObject* gameObject) {
+        return gameObject->AddComponent<SettingsScreenController>();
+    };
+}
+
+void SceneLoader::PreLoadScenes()
+{
+    std::vector<std::string> sceneNames = Config::GetValue("scenes", std::vector<std::string> {});
```

Comment:
is the second type here necessary? I am wondering if CTAD is smart enough to do it automagically.

```
std::vector<std::string> sceneNames = Config::GetValue("scenes", {});
```

(Total curiosity question, I don't know the answer)

Reply:
Apparently it is. I even tried something like this
`std::vector<const char*> sceneNames = Config::GetValue("scenes", {""});` with const char *
but still it couldn't quite get it.

Reply:
Yeah, I think that makes sense! Good data point, thanks for checking.